test/TorchSharpTest/TestTraining.cs

```diff
@@ -0,0 +1,516 @@
+
+                output.backward();
+
+                optimizer.step();
+            }
+        }
+
+        /// <summary>
+        /// Fully connected ReLU net with one hidden layer trained using Adagrad optimizer.
+        /// Taken from <see href="https://pytorch.org/tutorials/beginner/pytorch_with_examples.html#pytorch-optim"/>.
+        /// </summary>
+        [Fact]
+        public void TestTrainingAdagrad()
+        {
+            var lin1 = Linear(1000, 100);
+            var lin2 = Linear(100, 10);
+            var seq = Sequential(("lin1", lin1), ("relu1", ReLU()), ("lin2", lin2));
+
+            var x = Float32Tensor.randn(new long[] { 64, 1000 });
+            var y = Float32Tensor.randn(new long[] { 64, 10 });
+
+            double learning_rate = 0.00004f;
+            float prevLoss = float.MaxValue;
+            var optimizer = NN.Optimizer.Adagrad(seq.parameters(), learning_rate);
+            var loss = mse_loss(NN.Reduction.Sum);
+
+            for (int i = 0; i < 10; i++) {
+                var eval = seq.forward(x);
+                var output = loss(eval, y);
+                var lossVal = output.ToSingle();
+
+                Assert.True(lossVal < prevLoss);
+                prevLoss = lossVal;
+
+                optimizer.zero_grad();
+
+                output.backward();
+
+                optimizer.step();
+            }
+        }
+
+        /// <summary>
+        /// Fully connected ReLU net with one hidden layer trained using RMSprop optimizer.
+        /// Taken from <see href="https://pytorch.org/tutorials/beginner/pytorch_with_examples.html#pytorch-optim"/>.
+        /// </summary>
+        [Fact]
+        public void TestTrainingRMSLR()
+        {
+            var lin1 = Linear(1000, 100);
+            var lin2 = Linear(100, 10);
+            var seq = Sequential(("lin1", lin1), ("relu1", ReLU()), ("lin2", lin2));
+
+            var x = Float32Tensor.randn(new long[] { 64, 1000 });
+            var y = Float32Tensor.randn(new long[] { 64, 10 });
+
+            double learning_rate = 0.00004f;
+            float prevLoss = float.MaxValue;
+            var optimizer = NN.Optimizer.RMSProp(seq.parameters(), learning_rate);
+            var loss = mse_loss(NN.Reduction.Sum);
+
+            for (int i = 0; i < 10; i++) {
+                var eval = seq.forward(x);
+                var output = loss(eval, y);
+                var lossVal = output.ToSingle();
+
+                Assert.True(lossVal < prevLoss);
+                prevLoss = lossVal;
+
+                optimizer.zero_grad();
+
+                output.backward();
+
+                optimizer.step();
+            }
+        }
+
+        [Fact]
+        public void TestTrainingRMSAlpha()
+        {
+            var lin1 = Linear(1000, 100);
+            var lin2 = Linear(100, 10);
+            var seq = Sequential(("lin1", lin1), ("relu1", ReLU()), ("lin2", lin2));
+
+            var x = Float32Tensor.randn(new long[] { 64, 1000 });
```

Comment:
Just to mention as an aside that I'm not attached to this tensor creation API

DiffSharp uses this sort of thing:

    dsharp.tensor(obj, ...)

    dsharp.randn(..., dtype=Dtype.Float32)

It's quite possible we should support identical APIs or close to

    tsharp.tensor(obj, ...)

    tsharp.randn(..., dtype=....)

where replacing "torch.*" byt "tsharp" or "dsharp" gives fairly good results.